packages/next-swc/crates/core/src/modularize_imports.rs

```diff
@@ -0,0 +1,233 @@
+                    if self.config.prevent_full_import {
+                        panic!(
+                            "import {:?} causes the entire module to be imported",
+                            old_decl
+                        );
+                    } else {
+                        // Give up
+                        return vec![old_decl.clone()];
+                    }
+                }
+            }
+        }
+        out
+    }
+}
+
+impl FoldImports {
+    fn should_rewrite<'a>(&'a self, name: &'a str) -> Option<Rewriter<'a>> {
+        for (regex, config) in &self.packages {
+            let group = regex.captures(name);
+            if let Some(group) = group {
+                let group = group
+                    .iter()
+                    .map(|x| x.map(|x| x.as_str()).unwrap_or_default())
+                    .collect::<Vec<&str>>();
+                return Some(Rewriter {
+                    renderer: &self.renderer,
+                    key: name,
+                    config,
+                    group,
+                });
+            }
+        }
+        None
+    }
+}
+
+impl Fold for FoldImports {
+    noop_fold_type!();
+    fn fold_module(&mut self, mut module: Module) -> Module {
+        let mut new_items: Vec<ModuleItem> = vec![];
+        for item in module.body {
+            match item {
+                ModuleItem::ModuleDecl(ModuleDecl::Import(decl)) => {
+                    match self.should_rewrite(&decl.src.value) {
+                        Some(rewriter) => {
+                            let rewritten = rewriter.rewrite(&decl);
+                            new_items.extend(
+                                rewritten
+                                    .into_iter()
+                                    .map(|x| ModuleItem::ModuleDecl(ModuleDecl::Import(x))),
+                            );
+                        }
+                        None => new_items.push(ModuleItem::ModuleDecl(ModuleDecl::Import(decl))),
+                    }
+                }
+                x => {
+                    new_items.push(x);
+                }
+            }
+        }
+        module.body = new_items;
+        module
+    }
+}
+
+pub fn modularize_imports(config: Config) -> impl Fold {
+    let mut folder = FoldImports {
+        renderer: handlebars::Handlebars::new(),
+        packages: vec![],
+    };
+    folder
+        .renderer
+        .register_helper("lowerCase", Box::new(helper_lower_case));
+    folder
+        .renderer
+        .register_helper("upperCase", Box::new(helper_upper_case));
+    folder
+        .renderer
+        .register_helper("camelCase", Box::new(helper_camel_case));
+    for (mut k, v) in config.packages {
+        // XXX: Should we keep this hack?
+        if !k.starts_with('^') && !k.ends_with('$') {
+            k = format!("^{}$", k);
+        }
+        folder
+            .packages
+            .push((Regex::new(&k).expect("transform-imports: invalid regex"), v));
```

Comment:
This will be slow, because `Regex` will be created once per file.

Reply:
You can use https://docs.rs/swc_cached/0.1.1/swc_cached/regex/struct.CachedRegex.html

Reply:
Or use `once_cell::Lazy` https://github.com/vercel/next.js/blob/canary/packages/next-swc/crates/core/src/emotion/mod.rs#L75